### ISSUE TITLE
M04s

### ISSUE THREAD

**Issue by hermanneduard** [vivier/phomemo-tools#27]
Hi, i have a Phomemo M04s, thats seems to use not regular ESC/POS Commands. Some commands seem to be identical, but printing just text is (for me) not possible. I grabbed the Bluetooth-Commands but i was not able to reverse-engineer how to print a picture and how to encode the picture. Can you help?

**Comment by hermanneduard** [comment]
I connect via xcode/swift with my own routine. The connection is stable, if i send the previously grabbed packets, the printer prints them, but i (for my case) get it not working to generate the packets for an image or text… ;(

**Comment by vivier** [comment]
It seems Phomemo provides the linux driver for the M04s.
See https://phomemo.com/pages/drivers
I don't think files are specific to CentOS 7.

**Comment by Codeleiche** [comment]
I'm sorry, I know that it's off topic, but has anyone tested the drivers from phomemo for the M04s? I'm about to buy a unit and 300dpi seams great, but need it working with CUPS

**Comment by PixelAlarm** [comment]
I was not able to get it working in Ubuntu…

**Comment by hermanneduard** [comment]
I got to work, to generate an image in my ios app, and got it successfully printed on the M04s. I did it, by looking at the code for the chromeOs/chrome plugin (javascript). 
The main difference is, that the M04s uses kind of a lz1x compression.

**Comment by sbonn13** [comment]
@hermanneduard @PixelAlarm Do you use the phomemo M04s printer with Linux? Did you manage it finally? I would like to use it as a printer for my Raspberry Pi project, but I can buy it only if there is a chance to work.

**Comment by hermanneduard** [comment]
hi @Rajan1803 - i don't know how M02S works, but i made it work by compressing the image data as lz1 and generating the correct header. i made a bluetooth dump to find it out. Phomemo offers a Google Chrome extension, there you can find the models and header data, also the compression type as readable source code, this helped a lot to find out.

**Comment by Rajan-p-simform** [comment]
Hello @hermanneduard ,
Thank you so much for your response! I managed to get the printer working by reviewing the Chrome extension code you mentioned — that was really helpful.

However, I'm now facing an issue where the image prints in small segments. I'm sending the data in chunks, but when I send just 3 chunks, the printer takes around 15–20 pauses — printing a few lines, then pausing for a second before continuing. Eventually, the full image is printed correctly, but the process isn’t continuous like it is in the official Phomemo app.

I’m wondering if there’s something I might be missing? If you have any insights or tips, I’d really appreciate your help.

Thanks again!

**Comment by ooki1jp** [comment]
## M04AS Protocol Reverse-Engineering Results (BTSnoop HCI Log Analysis)

I've been working on M04AS support and wanted to share my findings from analyzing BTSnoop HCI logs captured while the official Phomemo Android app communicates with the printer. Hopefully this unblocks progress on M04s/M04AS support.

**TL;DR**: LZO compression is **not needed**. The official app itself sends raw uncompressed raster data. The protocol is relatively straightforward once you know the command sequence.

---

### BLE Service & Characteristics

| UUID | Type | Description |
| --- | --- | --- |
| `0000ff00-0000-1000-8000-00805f9b34fb` | Service | Main print service |
| `0000ff01-0000-1000-8000-00805f9b34fb` | Notify | Status notifications |
| `0000ff02-0000-1000-8000-00805f9b34fb` | Write | Data transmission |
| `0000ff03-0000-1000-8000-00805f9b34fb` | Notify | Additional notifications |

### Data Transfer

The official app uses **L2CAP CoC** (Connection-Oriented Channels) for fast transfer with credit-based flow control. However, standard **GATT writes work fine** as a fallback with these parameters:

| Parameter | Value |
| --- | --- |
| Chunk size | 205 bytes |
| Burst size | 3 chunks |
| Delay | 50ms per burst |
| Write mode | Without Response |

---

### Hardware

- **Print width**: 154 bytes = 1232 dots
- **Paper width**: 110mm
- **Resolution**: ~300 DPI (estimated from 1232 dots / 110mm)
- **Byte order**: Little-endian for multi-byte values

---

### Complete Command Sequence

The printer expects commands in this exact order:

```
1. Density
2. Heat/Speed parameter
3. Unknown init command
4. Compression mode
5. Raster header
6. Raster data
7. Paper feed (footer)
```

#### 1. Density Command

```
1F 11 02 DD    <- DD = density (0x00-0x0F, default 0x04)
```

#### 2. Heat/Speed Parameter (correlated with density)

```
1F 11 37 PP    <- PP = parameter (varies with density)
```

| App Setting | Density (cmd 0x02) | Param (cmd 0x37) |
| --- | --- | --- |
| Light | `0x01` | `0x64` (100) |
| Normal | `0x04` | `0x96` (150) |

Formula: `param = 100 + (density - 1) * 50 / 3`

#### 3. Unknown Init Command

```
1F 11 0B       <- No parameter. Required for proper operation.
```

#### 4. Compression Mode

```
1F 11 35 CC    <- 0x00 = Raw (OFF), 0x01 = LZO (ON)
```

**Key finding**: The official app sends `0x00` -- **LZO compression is not required.**

#### 5. Raster Header (ESC/POS GS v 0)

```
1D 76 30 00 WL WH HL HH
```

Example (154 bytes x 300 lines): `1D 76 30 00 9A 00 2C 01`

#### 6. Raster Data

Raw 1-bit monochrome. 1 = black, 0 = white, MSB first. Total = width_bytes x height.

#### 7. Footer (Paper Feed)

```
1B 64 02 1B 64 02    <- Feed 2 lines, sent twice
```

---

### Full Hex Example

```
1F 11 02 04              # Density (default)
1F 11 37 96              # Heat/speed (density=4 -> 0x96)
1F 11 0B                 # Init
1F 11 35 00              # Compression OFF
1D 76 30 00 9A 00 2C 01  # Header (154x300)
[46,200 bytes of raw bitmap data...]
1B 64 02 1B 64 02        # Paper feed
```

---

### Python: Image to Raster

```python
from PIL import Image

def image_to_raster(img: Image.Image, paper_width_bytes: int = 154) -> tuple[bytes, int]:
    paper_width_dots = paper_width_bytes * 8  # 1232 dots
    aspect = img.height / img.width
    target_width = min(img.width, paper_width_dots)
    target_height = int(target_width * aspect)
    img = img.resize((target_width, target_height))
    img = img.convert('1')  # 1-bit dithering

    width, height = img.width, img.height
    image_width_bytes = (width + 7) // 8

    padding = paper_width_bytes - image_width_bytes
    left_pad = padding // 2
    right_pad = padding - left_pad

    raster = bytearray()
    pixels = img.load()

    for y in range(height):
        raster.extend(bytes(left_pad))
        row = bytearray(image_width_bytes)
        for x in range(width):
            if pixels[x, y] == 0:  # Black pixel
                row[x // 8] |= (1 << (7 - (x % 8)))
        raster.extend(row)
        raster.extend(bytes(right_pad))

    return bytes(raster), height
```

---

### Open Questions

- **`1F 11 0B`** -- haven't determined its exact purpose, but omitting it causes issues.
- **Speed command** `1B 4E 0D` -- seen on other Phomemo models but M04AS doesn't appear to use it.
- **LZO mode** -- the protocol supports it, but since the official app doesn't use it and raw mode works, there's no real reason to implement it.

All findings captured by enabling BTSnoop HCI logging in Android developer options, printing at various density settings, and diffing the captures. Hope this helps!
